Sources/SourceKitLSP/SourceKitLSPServer.swift

```diff
@@ -594,11 +594,19 @@ public actor SourceKitLSPServer {
       // The latter might happen if there is an existing SwiftPM workspace that hasn't been reloaded after a new file
       // was added to it and thus currently doesn't know that it can handle that file. In that case, we shouldn't open
       // a new workspace for the same root. Instead, the existing workspace's build system needs to be reloaded.
-      if let workspace = await self.createWorkspace(WorkspaceFolder(uri: DocumentURI(url))),
-        await workspace.buildSystemManager.fileHandlingCapability(for: uri) == .handled,
-        let projectRoot = await workspace.buildSystemManager.projectRoot,
-        !projectRoots.contains(projectRoot)
-      {
+      let workspace = await self.createWorkspace(WorkspaceFolder(uri: DocumentURI(url))) { buildSystem in
+        guard let buildSystem else {
+          return false
+        }
+        do {
+          try await buildSystem.generateBuildGraph(allowFileSystemWrites: false)
+        } catch {
+          return false
+        }
+        let projectRoot = await buildSystem.projectRoot
+        return await buildSystem.fileHandlingCapability(for: uri) == .handled && !projectRoots.contains(projectRoot)
```

Comment:
We do reject `!projectRoots.contains(projectRoot)` before we create the workspace because it’s part of the `satisfying` closure passed to `createWorkspace`. 

We cannot perform this check before generating the build graph because we `projectRoot` is a property of `buildSystem`. And when you create a `SwiftPMBuildSystem`, for the `Sources` folder, I think it creates a build system with project root set to the folder that contains the `Package.swift`.

But we can hoist this check to be before the build graph generation, which makes sense.

Reply:
Just to clarify, I meant rejecting before the call to `createWorkspace`, based on the workspace URL. But yeah looks like the build system root can differ from the workspace.